runtime/gc_realtime/RealtimeAccessBarrier.cpp

```diff
@@ -549,12 +545,12 @@ MM_RealtimeAccessBarrier::jniGetStringCritical(J9VMThread* vmThread, jstring str
 	} else {
 		// acquire access and return a direct pointer
 #if defined(J9VM_INTERP_ATOMIC_FREE_JNI)
-		VM_VMAccess::inlineEnterVMFromJNI(vmThread);
-	  	vmThread->jniCriticalDirectCount += 1;
-		hasVMAccess = true;
-#else /* J9VM_INTERP_ATOMIC_FREE_JNI */
-		MM_JNICriticalRegion::enterCriticalRegion(vmThread);
+		if (!hasVMAccess) {
```

Comment:
Also just making a note for myself to come look at the hasVMAccess.  Not sure how one thread entering here could have it and another thread not have it.

Reply:
I suspect this code can be refactored/optimized at some point.